### PR TITLE
Subscription refactoring 4

### PR DESCRIPTION
**Pull request by federicocappelli**
Task/Issue URL: https://app.asana.com/0/1205842942115003/1206805455884775/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2961
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2871
What kind of version bump will this require?: Major

CC: @miasma13 

**Description**:

- Subscription API services and flows abstracted using protocols
- Mocks created for all protocols

**Mocks:**
APIServiceMock
AuthAPIServiceMock
StorePurchaseManagerMock
SubscriptionAPIServiceMock
AccountKeychainStorageMock
SubscriptionTokenKeychainStorageMock
SubscriptionUIHandlerMock
StripePurchaseFlowMock
AppStoreRestoreFlowMock
AppStorePurchaseFlowMock
AppStoreAccountManagementFlowMock

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Sources/SubscriptionTestingUtilities/AccountManagerMock.swift [hunk 1]
```diff
@@ -62,7 +61,7 @@ public final class AccountManagerMock: AccountManaging {
 
     }
 
-    public func hasEntitlement(for entitlement: Entitlement.ProductName, cachePolicy: CachePolicy) async -> Result<Bool, Error> {
+    public func hasEntitlement(forProductName entitlement: Entitlement.ProductName, cachePolicy: APICachePolicy) async -> Result<Bool, Error> {
```
[the remark this answers is not in the document]
Comment (federicocappelli) [reply Sources/SubscriptionTestingUtilities/AccountManagerMock.swift:64]: I disagree, The function name should be as clear as possible, the old version had a couple of problems: 1) the product name was called "entitlement", then the real type was an `Entitlement.ProductName` and not a `Entitlement` 2) no need to force the "user" to dig in the function param type in order to understand the function meaning when just a `hasEntitlement(forProductName` suffice

File: Sources/Subscription/Flows/AppStore/AppStoreAccountManagementFlow.swift [hunk 2]
```diff
@@ -20,36 +20,39 @@ import Foundation
...
+}
+
+@available(macOS 12.0, iOS 15.0, *)
+public protocol AppStoreAccountManagementFlowing {
+    @discardableResult func refreshAuthTokenIfNeeded() async -> Result<String, AppStoreAccountManagementFlowError>
+}
+
 @available(macOS 12.0, iOS 15.0, *)
-public final class AppStoreAccountManagementFlow {
+public final class AppStoreAccountManagementFlow: AppStoreAccountManagementFlowing {
```
[the remark this answers is not in the document]
Comment (federicocappelli) [reply Sources/Subscription/Flows/AppStore/AppStoreAccountManagementFlow.swift:34]: I'm trying with the second option

File: Sources/Subscription/SubscriptionManager.swift [hunk 3]
```diff
@@ -22,8 +22,8 @@ import Common
 public protocol SubscriptionManaging {
 
     var accountManager: AccountManaging { get }
-    var subscriptionService: SubscriptionService { get }
-    var authService: AuthService { get }
+    var subscriptionAPIService: SubscriptionAPIServicing { get }
+    var authAPIService: AuthAPIServicing { get }
```
[the remark this answers is not in the document]
Comment (federicocappelli) [reply Sources/Subscription/SubscriptionManager.swift]: I renamed Auth and Subscription to xxxEndpointService and removed the *ing suffix like the other class & protocols

**Comment by federicocappelli** [comment]
@miasma13  ready for a second review, all renaming should be complete

**Review by miasma13 — APPROVED** [review]

File: Sources/Subscription/API/SubscriptionEndpointService.swift [hunk 4]
```diff
@@ -111,42 +125,34 @@ public final class SubscriptionService: APIService {
...
-        public let email: String?
-        public let entitlements: [Entitlement]
-        public let subscription: Subscription
+extension DefaultSubscriptionEndpointService {
+
+    public init(currentServiceEnvironment: SubscriptionEnvironment.ServiceEnvironment) {
+        self.currentServiceEnvironment = currentServiceEnvironment
+        let baseURL = currentServiceEnvironment == .production ? URL(string: "https://subscriptions.duckduckgo.com/api")! : URL(string: "https://subscriptions-dev.duckduckgo.com/api")!
+        let session = URLSession(configuration: URLSessionConfiguration.ephemeral)
+        self.apiService = DefaultAPIService(baseURL: baseURL, session: session)
```
[the remark this answers is not in the document]
Comment (federicocappelli) [reply Sources/Subscription/API/SubscriptionEndpointService.swift]: moved

File: Sources/Subscription/API/AuthEndpointService.swift [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+extension DefaultAuthEndpointService {
+
+    public init(currentServiceEnvironment: SubscriptionEnvironment.ServiceEnvironment) {
+        self.currentServiceEnvironment = currentServiceEnvironment
+        let baseURL = currentServiceEnvironment == .production ? URL(string: "https://quack.duckduckgo.com/api/auth")! : URL(string: "https://quackdev.duckduckgo.com/api/auth")!
+        let session = URLSession(configuration: URLSessionConfiguration.ephemeral)
+        self.apiService = DefaultAPIService(baseURL: baseURL, session: session)
+    }
```
[the remark this answers is not in the document]
Comment (federicocappelli) [reply Sources/Subscription/API/AuthEndpointService.swift]: moved

File: Sources/Subscription/Managers/StorePurchaseManager.swift [hunk 6]
```diff
@@ -20,12 +20,40 @@ import Foundation
 import StoreKit
 import Common
 
+public enum StoreError: Error {
+    case failedVerification
+}
+
+public enum PurchaseManagerError: Error {
```
[the remark this answers is not in the document]
Comment (federicocappelli) [reply Sources/Subscription/Managers/StorePurchaseManager.swift]: correct, renamed

File: Sources/Subscription/API/SubscriptionEndpointService.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (federicocappelli) [reply Sources/Subscription/API/SubscriptionEndpointService.swift]: nope, leftover

**Review by miasma13 — APPROVED** [review]

**Pull request merged**